ext/openssl/openssl.c

```diff
@@ -5309,6 +5330,717 @@ PHP_FUNCTION(openssl_pkcs7_decrypt)
 
 /* }}} */
 
+/* {{{ CMS S/MIME functions taken from PKCS#7 functions */
+
+/* {{{ proto bool openssl_cms_verify(string filename, int flags [, string signerscerts [, array cainfo [, string extracerts [, string content [, string pk7]]]], string sigfile, int encoding])
+   Verifies that the data block is intact, the signer is who they say they are, and returns the CERTs of the signers */
+PHP_FUNCTION(openssl_cms_verify)
+{
+	X509_STORE * store = NULL;
+	zval * cainfo = NULL;
+	STACK_OF(X509) *signers= NULL;
+	STACK_OF(X509) *others = NULL;
+	CMS_ContentInfo * cms = NULL;
+	BIO * in = NULL, * datain = NULL, * dataout = NULL, * p7bout  = NULL;
+	BIO *certout = NULL, *sigbio = NULL;
+	zend_long flags = 0;
+	char * filename;
+	size_t filename_len;
+	char * extracerts = NULL;
+	size_t extracerts_len = 0;
+	char * signersfilename = NULL;
+	size_t signersfilename_len = 0;
+	char * datafilename = NULL;
+	size_t datafilename_len = 0;
+	char * p7bfilename = NULL;
+	size_t p7bfilename_len = 0;
+	char * sigfile = NULL;
+	size_t sigfile_len = 0;
+	zend_long encoding = ENCODING_SMIME;
+
+	RETVAL_FALSE;
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "pl|p!ap!p!p!p!l", &filename, &filename_len,
+				&flags, &signersfilename, &signersfilename_len, &cainfo,
+				&extracerts, &extracerts_len, &datafilename, &datafilename_len,
+				&p7bfilename, &p7bfilename_len,
+				&sigfile, &sigfile_len, &encoding) == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	in = BIO_new_file(filename, PHP_OPENSSL_BIO_MODE_R(flags));
+	if (in == NULL) {
+		php_openssl_store_errors();
+		goto clean_exit;
+	}
+	if (sigfile && (flags & CMS_DETACHED)) {
+		sigbio = BIO_new_file(sigfile, PHP_OPENSSL_BIO_MODE_R(flags));
+		if ( encoding == ENCODING_SMIME )  {
+			php_error_docref(NULL, E_WARNING,
+					 "Detached signatures not possible with S/MIME encoding");
+			goto clean_exit;
+		}
+	} else  {
+		sigbio = in;	/* non-detached signature */
+	}
+
+	switch (encoding) {
+		case ENCODING_PEM:
+			cms = PEM_read_bio_CMS(sigbio, NULL, 0, NULL);
+				datain = in;
+				break;
+			case ENCODING_DER:
+				cms = d2i_CMS_bio(sigbio, NULL);
+				datain = in;
+				break;
+			case ENCODING_SMIME:
+				cms = SMIME_read_CMS(sigbio, &datain);
+				break;
+			default:
+				php_error_docref(NULL, E_WARNING, "Unknown encoding");
+				goto clean_exit;
+	}
+	if (cms == NULL) {
+		php_openssl_store_errors();
+		goto clean_exit;
+	}
+	if ( encoding != ENCODING_SMIME && !(flags & CMS_DETACHED) ) {
+		datain = NULL; /* when not detached, don't pass a real BIO */
+	}
+
+	if (extracerts) {
+		others = php_openssl_load_all_certs_from_file(extracerts);
+		if (others == NULL) {
+			goto clean_exit;
+		}
+	}
+
+	store = php_openssl_setup_verify(cainfo);
+
+	if (!store) {
+		goto clean_exit;
+	}
+	if (php_openssl_open_base_dir_chk(filename)) {
```

Comment:
should this be done actually before it's passed to BIO?

Reply:
Yes.